geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RebalanceWhileCreatingRegionDistributedTest.java

```diff
@@ -0,0 +1,193 @@
+    int locatorPort = locator.getPort();
+    MemberVM server1 = cluster.startServerVM(1, locatorPort);
+    MemberVM server2 = cluster.startServerVM(2, locatorPort);
+    MemberVM accessor = cluster.startServerVM(4, locatorPort);
+
+    // Add DistributionMessageObserver
+    String regionName = testName.getMethodName();
+    Stream.of(server1, server2, accessor)
+        .forEach(server -> server.invoke(() -> addDistributionMessageObserver(regionName)));
+
+    // Create regions in each server
+    server1.invoke(() -> createRegion(regionName, RegionShortcut.PARTITION));
+    server2.invoke(() -> createRegion(regionName, RegionShortcut.PARTITION));
+
+    // Asynchronously wait to create the proxy region in the accessor
+    accessor.invokeAsync(() -> waitToCreateProxyRegion(regionName));
+
+    // Connect client1
+    ClientVM client1 =
+        cluster.startClientVM(5, c -> c.withServerConnection(server1.getPort(), server2.getPort()));
+
+    // Do puts
+    client1.invoke(() -> {
+      Region<Integer, Integer> region =
+          ClusterStartupRule.clientCacheRule.createProxyRegion(regionName);
+      IntStream.range(0, 3).forEach(i -> region.put(i, i));
+    });
+
+    // Start server3
+    MemberVM server3 = cluster.startServerVM(3, locatorPort);
+
+    // Create region in server3
+    server3.invoke(() -> createRegion(regionName, RegionShortcut.PARTITION));
+
+    // Add DistributionMessageObserver to server3
+    server3.invoke(() -> addDistributionMessageObserver(regionName));
+
+    // Rebalance
+    server1.invoke(() -> ClusterStartupRule.getCache().getResourceManager().createRebalanceFactory()
+        .start().getResults());
+
+    // Stop server3
+    server3.invoke(() -> ClusterStartupRule.getCache().close());
+
+    // Connect client to accessor
+    ClientVM client2 =
+        cluster.startClientVM(6, c -> c.withServerConnection(accessor.getPort())
+            .withCacheSetup(cf -> cf.setPoolReadTimeout(20000)));
+
+    // Do puts
+    client2.invoke(() -> {
+      Region<Integer, Integer> region =
+          ClusterStartupRule.clientCacheRule.createProxyRegion(regionName);
+      IntStream.range(0, 3).forEach(i -> region.put(i, i));
+    });
+  }
+
+  private void createRegion(String regionName, RegionShortcut shortcut) {
+    PartitionAttributesFactory<Integer, Integer> paf = new PartitionAttributesFactory<>();
+    paf.setRedundantCopies(0);
+    paf.setTotalNumBuckets(3);
+    if (shortcut.isProxy()) {
+      paf.setLocalMaxMemory(0);
+    }
+
+    RegionFactory<Integer, Integer> rf =
+        ClusterStartupRule.getCache().createRegionFactory(shortcut);
+    rf.setPartitionAttributes(paf.create());
+
+    rf.create(regionName);
+  }
+
+  private void waitToCreateProxyRegion(String regionName) throws Exception {
+    logger.info(
+        "RebalanceWhileCreatingRegionDistributedTest.waitToCreateRegion about to wait for Before_RemoveBucketMessage gate");
+    blackboard.waitForGate("Before_RemoveBucketMessage");
```

Comment:
Could the gate names in this test be extracted to constants please? It makes following the flow of events through the test easier if you can see each place a constant is used vs. having to search for a specific String.